controller/rollout_save_controller.go

```diff
@@ -0,0 +1,1091 @@
+	err := util.AddFinalizer(longhornFinalizerKey, rolloutSave)
+	if err != nil {
+		return rolloutSave, err
+	}
+
+	rolloutSave, err = c.ds.UpdateRolloutSave(rolloutSave)
+	if err != nil {
+		return rolloutSave, err
+	}
+
+	rolloutSave.Status.State = longhorn.RolloutSaveStateGenerate
+	rolloutSave.Status.Version = meta.Version
+	rolloutSave.Status.GitCommit = meta.GitCommit
+
+	return rolloutSave, nil
+}
+
+func (c *RolloutSaveController) UploadRollout(rolloutSave *longhorn.RolloutSave, archievePath, tempDir string, backupTargetClient engineapi.BackupTargetClientRolloutInterface) {
+	log := getLoggerForRolloutSave(c.logger, rolloutSave.Name, backupTargetClient.GetURL())
+
+	var err error
+	var errConditionMessage string
+	defer func() {
+		if err == nil {
+			rolloutConfig, err := backupTargetClient.RolloutGetConfig(rolloutSave.Name, rolloutSave.Status.Version)
+			if err != nil {
+				errConditionMessage = RolloutSaveErrGetConfigFailed
+			} else {
+				rolloutSave.Status.State = longhorn.RolloutSaveStateReady
+				rolloutSave.Status.ManagerImage = rolloutConfig.ManagerImage
+				rolloutSave.Status.CreatedAt = metav1.Time{Time: time.Now().UTC()}
+
+				log.Info(RolloutSaveMsgUploadBackupTarget)
+				c.eventRecorder.Eventf(rolloutSave, corev1.EventTypeNormal, EventReasonUploaded, RolloutSaveMsgUploadBackupTarget)
+			}
+		}
+
+		if err != nil {
+			c.rolloutSaveErrorHandle(rolloutSave, longhorn.RolloutSaveConditionReasonUpload, errConditionMessage, err, log)
+		}
+
+		_, err = c.ds.UpdateRolloutSaveStatus(rolloutSave)
+		if err != nil && !apierrors.IsConflict(errors.Cause(err)) {
+			log.WithError(err).Debug(RolloutSaveMsgRequeueOnError)
+			c.enqueueRolloutSave(rolloutSave)
+		}
+	}()
+
+	defaultEngineImage, err := c.ds.GetSettingValueExisted(types.SettingNameDefaultEngineImage)
+	if err != nil {
+		errConditionMessage = fmt.Sprintf(RolloutSaveErrGetFailedFmt, "default engine image")
+		return
+	}
+
+	timeout := 60 * time.Second
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			err = errors.Wrap(err, "timeout uploading rollout")
+			return
+		case <-ticker.C:
+			var rolloutConfig *bsRollout.RolloutConfig
+			rolloutConfig, err = backupTargetClient.RolloutGetConfig(rolloutSave.Name, rolloutSave.Status.Version)
+			if err != nil && !types.ErrorIsNotFound(err) {
+				errConditionMessage = RolloutSaveErrGetConfigFailed
+				continue
+			}
+
+			if rolloutConfig != nil {
+				return
+			}
+
+			_, err = backupTargetClient.RolloutUpload(rolloutSave.Name, archievePath, rolloutSave.Status.Version, rolloutSave.Status.GitCommit, c.managerImage, defaultEngineImage)
+			if err != nil {
+				if types.ErrorAlreadyExists(err) {
+					err = nil
+					continue
+				}
+
+				log.WithError(err).Debugf("retry uploading %v to backup target", archievePath)
+				continue
+			}
+
```

Comment:
At success load, this will re-loop at the next ticker.c and return if the rollout.cfg exists in the backup target.

I am a little confused, can you elaborate on how this function will get trapped into the timeout error?

Reply:
Let's say the execution of `backupTargetClient.RolloutUpload` takes 70s, which is larger than the timeout. Then in the next loop, could the function get into `case <-timer.C`?

Reply:
1. There should be sufficient time for the upload, `60s` is definitely too short as previously raised in https://github.com/longhorn/longhorn-manager/pull/1489#discussion_r979560333

2. The intention is to allow retries for the upload. So when the retry exceeds the timeout, the logic goes to `case <-timer.C`.
  1. The `err`  will be wrapped with `RolloutSaveErrTimeout`.
  2. return method.
  3. Post method `defer func()` updates RolloutSave status. Here the `err` will be updated to the `rolloutSave.Status.Conditions`.
  ```
  "rollout-save upload file failed": {
			state:                       longhorn.RolloutSaveStateUpload,
			expectState:                 longhorn.RolloutSaveStateError,
			expectErrorConditionMessage: fmt.Sprintf("%v: %v:", RolloutSaveErrUploadFailed, RolloutSaveErrTimeout),
		},
  ```

Reply:
I mean, when the upload **succeeds** but the time it takes is longer than the timeout value, should we error out?

Reply:
I see what you mean. Updated.

1. Upload taking longer than the timeout.
1. Upload succeeded.
1. For loop goes to `case <- timer.c` and here the `err` remains `nil` after `errors.Wrap`.
1. `case <- timer.c` returns.
1. defer func() handles updates RolloutSave status to ready because `err == nil`.

```
// ex: unit-test
"rollout-save upload file exceed timeout": {
			rolloutSaveName: TestRolloutSaveUploadExceedTimeout,
			state:           longhorn.RolloutSaveStateUpload,
			expectState:     longhorn.RolloutSaveStateReady,
		},
```

Reply:
I see. Then this should work fine.